ros/src/actuation/vehicles/packages/ymc/node/g30esli_interface/g30esli_ros.cpp

```diff
@@ -0,0 +1,329 @@
+  current_twist_.header.frame_id = "base_link";
+  current_twist_.header.stamp = now;
+  current_twist_.twist.linear.x = lv;
+  current_twist_.twist.angular.z = az;
+
+  // update vehicle status
+  vehicle_status_.header = current_twist_.header;
+
+  // drive/steeringmode
+  if (status_.status.mode == G30ESLI_MODE_MANUAL)
+  {
+    vehicle_status_.drivemode = autoware_msgs::VehicleStatus::MODE_MANUAL;
+    vehicle_status_.steeringmode = autoware_msgs::VehicleStatus::MODE_MANUAL;
+  }
+  else if (status_.status.mode == G30ESLI_MODE_AUTO)
+  {
+    vehicle_status_.drivemode = autoware_msgs::VehicleStatus::MODE_AUTO;
+    vehicle_status_.steeringmode = autoware_msgs::VehicleStatus::MODE_AUTO;
+  }
+
+  // gearshift
+  if (status_.status.shift == G30ESLI_SHIFT_DRIVE)
+  {
+    vehicle_status_.gearshift = 1;
+  }
+  else if (status_.status.shift == G30ESLI_SHIFT_REVERSE)
+  {
+    vehicle_status_.gearshift = 2;
+  }
+  else if (status_.status.shift == G30ESLI_SHIFT_NEUTRAL)
+  {
+    vehicle_status_.gearshift = 4;
+  }
+
+  // speed
+  vehicle_status_.speed = status_.status.speed.actual;  // [kmph]
+
+  // drivepedal
+  vehicle_status_.drivepedal = status_.status.override.accel;  // [-]
+
+  // brakepedal
+  vehicle_status_.brakepedal = status_.status.override.brake;  // [-]
+
+  // angle
+  vehicle_status_.angle = -status_.status.steer.actual - steering_offset_deg;  // [deg]
+
+  // lamp
+  if (status_.status.override.flasher == G30ESLI_FLASHER_NONE)
+  {
+    vehicle_status_.lamp = 0;
+  }
+  else if (status_.status.override.flasher == G30ESLI_FLASHER_RIGHT)
+  {
+    vehicle_status_.lamp = autoware_msgs::VehicleStatus::LAMP_RIGHT;
+  }
+  else if (status_.status.override.flasher == G30ESLI_FLASHER_LEFT)
+  {
+    vehicle_status_.lamp = autoware_msgs::VehicleStatus::LAMP_LEFT;
+  }
+  else if (status_.status.override.flasher == G30ESLI_FLASHER_HAZARD)
+  {
+    vehicle_status_.lamp = autoware_msgs::VehicleStatus::LAMP_HAZARD;
+  }
+
+  // light
+  vehicle_status_.light = 0;  // not used
+}
+
+bool G30esliROS::checkOverride()
+{
+  return (status_.status.override.accel == 1 || status_.status.override.brake == 1);
+}
+
+bool G30esliROS::checkTimeout(const MODE& mode, const double& timeout)
+{
+  Command& cmd = commands_[(int)mode];
+  double dtms = (ros::Time::now() - cmd.time).toSec() * 1000;
+  bool timeouted = (dtms > timeout);
+  return (cmd.received && timeouted);
+}
+
+void G30esliROS::checkRestart(const MODE& mode)
+{
+  static bool initialize = false;
+  static double stop_keep_dt = 0.0;
+  static ros::Time prev_time, start_reset_time, end_reset_time;
+
+  ros::Time curr_time = ros::Time::now();
+
+  if (!initialize)
+  {
+    prev_time = curr_time;
+    initialize = true;
+    return;
+  }
+
+  Command& cmd = commands_[(int)mode];
+  bool is_positive_command = (cmd.command.speed > 0.0);
+  bool is_stopped = (vehicle_status_.speed < 0.01);
+
+  if (!reset_command_)
+  {
+    if (!((curr_time - end_reset_time).toSec() > disable_reset_secs_))
```

Comment:
I think that "!" is unnecessary if the purpose is to wait for not entering the stop function for a certain period.

Reply:
It's bug. Fixed.